Scoped component is being attempted to be created from root scope in 4.2.0

**Describe the bug**
In versions prior 4.2.0 correctly scoped components are correctly loaded from scope upon which declare was called. 

In 4.2.0 same component is tried to be created from `_root_` scope. 

```
Caused by: org.koin.core.error.NoDefinitionFoundException: No definition found for type 'net.my.app.MyClass' on scope '['_root_']'. Check your Modules configuration and add missing type and/or qualifier! (Fix with AI)
                 	at org.koin.core.resolution.CoreResolverV2.resolveFromContext(CoreResolverV2.kt:160)
                 	at org.koin.core.scope.Scope.resolveFromContext(Scope.kt:329)
                 	at org.koin.core.scope.Scope.stackParametersCall(Scope.kt:292)
                 	at org.koin.core.scope.Scope.resolveInstance(Scope.kt:278)
                 	at org.koin.core.scope.Scope.resolve(Scope.kt:251)
```

The class is correctly scoped by dsl and it works fine in versions prior 4.2.0 

**To Reproduce**
Steps to reproduce the behavior:
1. Scope a class 
2. Call `getOrCreateScope("someId", named("MyScope")`, call `declare("someId")`
3. Upgrade to 4.2.0
4. Run 

**Expected behavior**
A compont should be loaded from correct scope instead from `_root_` scope
**Koin module and version:**
4.2.0

Unfortunately can't provide snippet.

--- a/projects/core/koin-core/src/commonMain/kotlin/org/koin/core/resolution/CoreResolverV2.kt
+++ b/projects/core/koin-core/src/commonMain/kotlin/org/koin/core/resolution/CoreResolverV2.kt
@@ -21,6 +21,7 @@ import org.koin.core.annotation.KoinInternalApi
 import org.koin.core.error.NoDefinitionFoundException
 import org.koin.core.instance.InstanceFactory
 import org.koin.core.instance.ResolutionContext
+import org.koin.core.instance.SingleInstanceFactory
 import org.koin.core.scope.Scope
 import org.koin.ext.getFullName
 
@@ -86,6 +87,17 @@ class CoreResolverV2(
             // 1. Registry on this linked scope
             val factory = findDefinitionInScope(linkedScope, ctx)
             if (factory != null) {
+                // #2379: a root FACTORY requested from a child scope keeps the ORIGINATING
+                // context, so its transitive (possibly scoped/archetype) dependencies
+                // resolve back in the requesting scope instead of falling through to _root_.
+                // The origin ctx already carries the scope + scopeArchetype and its params
+                // are already stacked on it, so no context switch or re-stacking is needed.
+                // Root SINGLES are excluded: a singleton must resolve its dependencies once,
+                // from root, and must never capture a scope-local instance (guard: #2325).
+                if (linkedScope.isRoot && factory !is SingleInstanceFactory<*>) {
+                    return factory.get(ctx) as T?
+                }
+
                 // we will loose parameters from parent context
                 val newCtx = ctx.newContextForScope(linkedScope)
                 if (linkedScope.scopeArchetype != null && !linkedScope.isRoot) {

--- a/projects/core/koin-core/src/commonTest/kotlin/org/koin/core/ArchetypeDeclareResolveTest.kt
+++ b/projects/core/koin-core/src/commonTest/kotlin/org/koin/core/ArchetypeDeclareResolveTest.kt
@@ -1,0 +1,95 @@
+package org.koin.core
+
+import org.koin.Simple
+import org.koin.core.annotation.KoinInternalApi
+import org.koin.core.component.getScopeId
+import org.koin.core.logger.Level
+import org.koin.core.module.KoinDslMarker
+import org.koin.core.module.Module
+import org.koin.core.qualifier.TypeQualifier
+import org.koin.dsl.ScopeDSL
+import org.koin.dsl.koinApplication
+import org.koin.dsl.module
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertNotNull
+
+/**
+ * Regression repro for issue #2379 (archetype-scope variant).
+ *
+ * The plain named-scope case is covered by DeclareInstanceTest. The case still
+ * reported failing in 4.2.0/4.2.1 (dees91) uses an *archetype* scope
+ * (activityRetainedScope): a scoped definition resolved through its archetype,
+ * whose transitive get() for a declared dependency falls through to _root_.
+ */
+class ArchetypeDeclareResolveTest {
+
+    open class Archetype
+    class ArchetypeExt : Archetype()
+
+    @KoinDslMarker
+    fun Module.scopeArchetype(scopeSet: ScopeDSL.() -> Unit) {
+        val qualifier = TypeQualifier(Archetype::class)
+        ScopeDSL(qualifier, this).apply(scopeSet)
+    }
+
+    @OptIn(KoinInternalApi::class)
+    @Test
+    fun archetype_scope_resolves_declared_transitive_dependency_issue_2379() {
+        val scopeModule = module {
+            scopeArchetype {
+                scoped { Simple.ComponentB(get()) }
+            }
+        }
+        val koin = koinApplication {
+            printLogger(Level.DEBUG)
+            modules(scopeModule)
+        }.koin
+
+        val archetypeExt = ArchetypeExt()
+        val scope = koin.createScope<ArchetypeExt>(
+            archetypeExt.getScopeId(), archetypeExt,
+            TypeQualifier(Archetype::class)
+        )
+        scope.declare(Simple.ComponentA())
+
+        // declared dependency must resolve from the scope it was declared in
+        val a = scope.getOrNull<Simple.ComponentA>()
+        assertNotNull(a)
+
+        // scoped ComponentB's transitive get() for ComponentA must resolve in-scope, not _root_
+        val b = scope.get<Simple.ComponentB>()
+        assertEquals(a, b.a)
+    }
+
+    // dees91 minimal shape: a ROOT factory depends on a SCOPED dependency,
+    // resolved transitively from the scope. v2 switches context to root when it
+    // finds the root factory via linked scopes, then can't resolve the scoped dep.
+    class Connector
+    class Interactor(val connector: Connector)
+
+    @OptIn(KoinInternalApi::class)
+    @Test
+    fun root_factory_depending_on_scoped_dep_resolves_in_scope_issue_2379() {
+        val scopeModule = module {
+            scopeArchetype {
+                scoped { Connector() }
+            }
+            factory { Interactor(get()) } // ROOT factory needs the scoped Connector
+        }
+        val koin = koinApplication {
+            printLogger(Level.DEBUG)
+            modules(scopeModule)
+        }.koin
+
+        val archetypeExt = ArchetypeExt()
+        val scope = koin.createScope<ArchetypeExt>(
+            archetypeExt.getScopeId(), archetypeExt,
+            TypeQualifier(Archetype::class)
+        )
+
+        // resolving the root factory FROM the scope must let its scoped dep resolve in-scope
+        val interactor = scope.get<Interactor>()
+        assertNotNull(interactor.connector)
+    }
+}

